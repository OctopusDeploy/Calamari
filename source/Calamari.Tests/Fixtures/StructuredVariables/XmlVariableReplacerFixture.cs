using System;
using System.Runtime.CompilerServices;
using Assent;
using Calamari.Common.Features.StructuredVariables;
using Calamari.Common.Plumbing.FileSystem;
using Calamari.Common.Plumbing.Variables;
using Calamari.Tests.Helpers;
using NUnit.Framework;

namespace Calamari.Tests.Fixtures.StructuredVariables
{
    [TestFixture]
    public class XmlVariableReplacerFixture : VariableReplacerFixture
    {
        public XmlVariableReplacerFixture()
            : base(new XmlFormatVariableReplacer(CalamariPhysicalFileSystem.GetPhysicalFileSystem(), new InMemoryLog()))
        {
        }

        [Test]
        public void DoesNothingIfThereAreNoVariables()
        {
            var vars = new CalamariVariables();
            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceAComment()
        {
            var vars = new CalamariVariables
            {
                { "/document/comment()", "New Comment" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceAnAttribute()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting[@id='id-1']/@id", "id-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceATextNode()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting[@id='id-1']/text()", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceAnElementsText()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting[@id='id-1']", "value<new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanInsertTextIntoAnEmptyElement()
        {
            var vars = new CalamariVariables
            {
                { "/document/empty", "value<new" }
            };

            RunTest(vars, "elements.xml");
        }

        [Test]
        public void CanInsertTextIntoASelfClosingElement()
        {
            var vars = new CalamariVariables
            {
                { "/document/selfclosing", "value<new" }
            };

            RunTest(vars, "elements.xml");
        }

        [Test]
        public void CanReplaceAnElementsChildrenWhenTheElementHasMixedContent()
        {
            var vars = new CalamariVariables
            {
                { "/document/mixed", "<newElement />" }
            };

            RunTest(vars, "elements.xml");
        }

        [Test]
        public void CanReplaceAnElementsChildren()
        {
            var vars = new CalamariVariables
            {
                { "//moreSettings", "<a /><b />" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceAnElementsChildrenWhenTheNewChildrenHaveNamespaces()
        {
            var vars = new CalamariVariables
            {
                { "//moreSettings", "<db:a />" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void DoesNotModifyDocumentWhenVariableCannotBeParsedAsMarkup()
        {
            var vars = new CalamariVariables
            {
                { "//moreSettings", "<<<<" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void DoesntTreatVariableAsMarkupWhenReplacingAnElementThatContainsNoElementChildren()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting[@id='id-1']", "<a />" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceMultipleElements()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceMultipleElementsInDifferentPartsOfTheTree()
        {
            var vars = new CalamariVariables
            {
                { "//setting", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceCData()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting[@id='id-3']/text()", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceCDataInAnElement()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting[@id='id-3']", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceProcessingInstructions()
        {
            var vars = new CalamariVariables
            {
                { "/document/processing-instruction('xml-stylesheet')", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanInferNamespacePrefixesFromDocument()
        {
            var vars = new CalamariVariables
            {
                { "/document/unique:anotherSetting", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanUseXPath2Wildcards()
        {
            var vars = new CalamariVariables
            {
                { "//*:node", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void DoesNotThrowOnUnrecognisedNamespace()
        {
            var vars = new CalamariVariables
            {
                { "//env:something", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void UsesTheFirstNamespaceWhenADuplicatePrefixIsFound()
        {
            var vars = new CalamariVariables
            {
                { "//parent/dupe:node", "value-new" }
            };

            RunTest(vars, "duplicate-prefixes.xml");
        }

        [Test]
        public void ShouldIgnoreOctopusPrefix()
        {
            var vars = new CalamariVariables
            {
                { "Octopus.Release.Id", "999" }
            };
            RunTest(vars, "ignore-octopus.xml");
        }

        [Test]
        public void HandlesVariablesThatReferenceOtherVariables()
        {
            var vars = new CalamariVariables
            {
                { "/document/selfclosing", "#{a}" },
                { "a", "#{b}" },
                { "b", "value-new" },
            };
            RunTest(vars, "elements.xml");
        }

        [Test]
        public void ShouldPreserveEncodingUtf8DosBom()
        {
            this.Assent(ReplaceToHex(new CalamariVariables(), "enc-utf8-dos-bom.xml"),
                        TestEnvironment.AssentConfiguration);
        }

        [Test]
        public void ShouldPreserveEncodingUtf8UnixNoBom()
        {
            this.Assent(ReplaceToHex(new CalamariVariables(), "enc-utf8-unix-nobom.xml"),
                        TestEnvironment.AssentConfiguration);
        }

        [Test]
        public void ShouldPreserveEncodingUtf16DosBom()
        {
            this.Assent(ReplaceToHex(new CalamariVariables(), "enc-utf16-dos-bom.xml"),
                        TestEnvironment.AssentConfiguration);
        }

        [Test]
        public void ShouldPreserveEncodingWindows1252DosNoBom()
        {
            this.Assent(ReplaceToHex(new CalamariVariables(), "enc-windows1252-dos-nobom.xml"),
                        TestEnvironment.AssentConfiguration);
        }

        void RunTest(CalamariVariables vars,
                     string file,
                     [CallerMemberName]
                     string testName = null,
                     [CallerFilePath]
                     string filePath = null)
        {
            // ReSharper disable once ExplicitCallerInfoArgument
            this.Assent(Replace(vars, file), TestEnvironment.AssentXmlConfiguration, testName, filePath);
        }
    }
}