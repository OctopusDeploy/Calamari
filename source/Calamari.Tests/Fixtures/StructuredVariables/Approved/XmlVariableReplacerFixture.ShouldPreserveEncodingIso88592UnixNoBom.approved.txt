3C 3F 78 6D 6C 20 76 65   <?xml ve
72 73 69 6F 6E 3D 22 31   rsion="1
2E 30 22 20 65 6E 63 6F   .0" enco
64 69 6E 67 3D 22 69 73   ding="is
6F 2D 38 38 35 39 2D 32   o-8859-2
22 3F 3E 0A 3C 74 6F 70   "?>^<top
2D 6E 61 6D 65 73 2D 66   -names-f
6F 72 2D 62 6F 79 73 3E   or-boys>
0A 20 20 3C 6E 61 6D 65   ^  <name
3E 4A 61 6B 75 62 3C 2F   >Jakub</
6E 61 6D 65 3E 0A 20 20   name>^  
3C 6E 61 6D 65 3E 53 7A   <name>Sz
79 6D 6F 6E 3C 2F 6E 61   ymon</na
6D 65 3E 0A 20 20 3C 6E   me>^  <n
61 6D 65 3E 4B 61 63 70   ame>Kacp
65 72 3C 2F 6E 61 6D 65   er</name
3E 0A 20 20 3C 6E 61 6D   >^  <nam
65 3E 46 69 6C 69 70 3C   e>Filip<
2F 6E 61 6D 65 3E 0A 20   /name>^ 
20 3C 6E 61 6D 65 3E 4D    <name>M
69 63 68 61 B3 2B 3C 2F   icha?+</
6E 61 6D 65 3E 0A 20 20   name>^  
3C 6E 61 6D 65 3E 4D 61   <name>Ma
74 65 75 73 7A 3C 2F 6E   teusz</n
61 6D 65 3E 0A 20 20 3C   ame>^  <
6E 61 6D 65 3E 42 61 72   name>Bar
74 6F 73 7A 3C 2F 6E 61   tosz</na
6D 65 3E 0A 20 20 3C 6E   me>^  <n
61 6D 65 3E 57 6F 6A 63   ame>Wojc
69 65 63 68 3C 2F 6E 61   iech</na
6D 65 3E 0A 20 20 3C 6E   me>^  <n
61 6D 65 3E 41 64 61 6D   ame>Adam
3C 2F 6E 61 6D 65 3E 0A   </name>^
20 20 3C 6E 61 6D 65 3E     <name>
57 69 6B 74 6F 72 3C 2F   Wiktor</
6E 61 6D 65 3E 0A 20 20   name>^  
3C 6E 61 6D 65 3E 50 69   <name>Pi
6F 74 72 3C 2F 6E 61 6D   otr</nam
65 3E 0A 20 20 3C 6E 61   e>^  <na
6D 65 3E 4A 61 6E 3C 2F   me>Jan</
6E 61 6D 65 3E 0A 20 20   name>^  
3C 6E 61 6D 65 3E 44 61   <name>Da
77 69 64 3C 2F 6E 61 6D   wid</nam
65 3E 0A 20 20 3C 6E 61   e>^  <na
6D 65 3E 4D 69 6B 6F B3   me>Miko?
61 6A 3C 2F 6E 61 6D 65   aj</name
3E 0A 3C 2F 74 6F 70 2D   >^</top-
6E 61 6D 65 73 2D 66 6F   names-fo
72 2D 62 6F 79 73 3E      r-boys>